src/RenderWebGL.js

```diff
@@ -243,16 +243,45 @@ RenderWebGL.prototype.getSkinSize = function (drawableID) {
  * @returns {Boolean} True iff the Drawable is touching the color.
  */
 RenderWebGL.prototype.isTouchingColor = function(drawableID, color3b, mask3b) {
+    const gl = this._gl;
+    twgl.bindFramebufferInfo(gl, this._queryBufferInfo);
 
-    var gl = this._gl;
+    const drawable = Drawable.getDrawableByID(drawableID);
+    const bounds = drawable.getFastBounds();
 
-    twgl.bindFramebufferInfo(gl, this._queryBufferInfo);
+    // Limit queries to the stage size.
+    bounds.clamp(this._xLeft, this._xRight, this._yBottom, this._yTop);
+
+    // Use integer coordinates for queries - weird things happen
+    // when you provide float width/heights to gl.viewport and projection.
+    bounds.ceil();
 
-    // TODO: restrict to only the area overlapped by the target Drawable
-    // - limit size of viewport to the AABB around the target Drawable
-    // - draw only the Drawables which could overlap the target Drawable
-    // - read only the pixels in the AABB around the target Drawable
-    gl.viewport(0, 0, this._nativeSize[0], this._nativeSize[1]);
+    if (bounds.width == 0 || bounds.height == 0) {
+        // No space to query.
+        return false;
+    }
+
+    // Filter candidates by rough bounding box intersection.
+    // Do this before _drawThese, so we can prevent any GL operations
+    // and readback by returning early.
+    const candidateIDs = this._drawables.filter(function (testID) {
+        if (testID == drawableID) return false;
+        // Only draw items which could possibly overlap target Drawable.
+        let candidate = Drawable.getDrawableByID(testID);
+        let candidateBounds = candidate.getFastBounds();
+        return bounds.intersects(candidateBounds);
+    });
+
+    if (candidateIDs.length == 0) {
+        // No possible intersections based on bounding boxes.
+        return false;
+    }
+
+    // Limit size of viewport to the bounds around the target Drawable,
+    // and create the projection matrix for the draw.
+    gl.viewport(0, 0, bounds.width, bounds.height);
+    const projection = twgl.m4.ortho(
+        bounds.left, bounds.right, bounds.bottom, bounds.top, -1, 1);
```

Comment:
This big chunk of code is nearly identical in `isTouchingColor` and `isTouchingDrawables`... could it be pulled out into a helper function?